### DISCUSSION TITLE
How to name x&y axis after to_hist a TH2F histogram

### DISCUSSION THREAD

**Discussion by ZhenxuanZhang-Jensen** [scikit-hep/hist#342] [Q&A]
Suppose I just created a TH2F histogram and use to_hist() to make it a hist, and how can I name the x and y axis to different name?
```python
import random
import uproot
from ROOT import TFile, TH2F

of = TFile("test_hist.root","RECREATE")
h1 = TH2F("h1","Histogram",100,-5,5,100,-5,5)
for i in range(100):
    x = random.uniform(-4.9,4.9)
    y = random.uniform(-4.9,4.9)
    h1.Fill(x,y)
of.Write()
of.Close()

hist = uproot.open("test_hist.root")
hist["h1"].to_hist().plot2d_full()
```

**Comment by henryiii** [comment]
You need to rename both at the same time, since that's the only way to verify the names are valid. Just do this:

```python
h = hist["h1"].to_hist()
h.axes.name = ("x", "y")
h.plot2d_full()
```

**Reply by ZhenxuanZhang-Jensen** [reply]
Thank you for your reply. Still get some issue here:

```python
of = TFile("tmp_TwoD.root","RECREATE")
h1 = TH2F("h1","Histrogram",100,-5,5,100,-5,5)
array_GEN_H1_eta = events["Events"]["GEN_H1_eta"].arrays()[:]["GEN_H1_eta"]
array_GEN_H2_eta = events["Events"]["GEN_H2_eta"].arrays()[:]["GEN_H2_eta"]
for i in range(len(array_GEN_H1_eta)):
    h1.Fill(array_GEN_H1_eta[i],array_GEN_H2_eta[i])
of.Write()
of.Close()
tmp_TwoD = uproot.open("tmp_TwoD.root")
histTwoD = tmp_TwoD["h1"].to_hist()
histTwoD.axes.name = ("x","y")
histTwoD.plot2d_full()
```
the error is AttributeError: can't set attribute

**Reply by henryiii** [reply]
What version of Hist are you using?

**Reply by ZhenxuanZhang-Jensen** [reply]
2.5.1, I believe

**Reply by ZhenxuanZhang-Jensen** [reply]
Henryiii, Do you have any solutions soon? Should I change the version or do anything else?

**Reply by LovelyBuggies** [reply]
@ZhenxuanZhang-Jensen I think this issue should be solved in #288 and be released in 2.5.0. Does your histogram `histTwoD` have differences with this?

```python
h = Hist.new.Regular(50, -3, 3, name="X").Regular(20, -3, 3, name="Y").Double()
h.axes.name = ("A", "B")

assert h.axes[0].name == "A"
assert h.axes[1].name == "B"
```

**Reply by henryiii** [reply]
I had to rebuild ROOT (brew ROOT and conda root both do not support Xcode 13.1 yet unless you rebuild), but it works fine for me:

```python
import random
from ROOT import File, TH2F

of = TFile("test_hist.root","RECREATE")
h1 = TH2F("h1","Histogram",100,-5,5,100,-5,5)
for i in range(100):
    x = random.uniform(-4.9,4.9)
    y = random.uniform(-4.9,4.9)
    h1.Fill(x,y)
of.Write()
of.Close()

import uproot
from hist import Hist
hist = uproot.open("test_hist.root")
h = hist["h1"].to_hist()
h.axes.name = ("x", "y")
h.plot2d_full()
```

![Figure_1](https://user-images.githubusercontent.com/4616906/141923567-2b7859c0-2b9b-4f64-bb57-9a190642cc3e.png)

**Reply by ZhenxuanZhang-Jensen** [reply]
okay...So this is the version problem. I work with lxplus CERN server, so maybe some version problem happen...

**Reply by ZhenxuanZhang-Jensen** [reply]
> @ZhenxuanZhang-Jensen I think this issue should be solved in #288 and be released in 2.5.0. Does your histogram `histTwoD` have differences with this?
> 
> ```python
> h = Hist.new.Regular(50, -3, 3, name="X").Regular(20, -3, 3, name="Y").Double()
> h.axes.name = ("A", "B")
> 
> assert h.axes[0].name == "A"
> assert h.axes[1].name == "B"
> ```

Yes, when I run this, it also meets some bugs like AttributeError: can't set attribute

**Reply by henryiii** [reply]
What is:

```python
import hist
print(hist.__version__)
```

Also can you update Hist?

**Reply by ZhenxuanZhang-Jensen** [reply]
is "2.5.1". I don't think I can update Hist but maybe I can find a newer source of python environment to use in lxplus.

**Reply by henryiii** [reply]
How are you setting up Python on lxplus?

**Reply by ZhenxuanZhang-Jensen** [reply]
I use `source /cvmfs/sft.cern.ch/lcg/views/dev4cuda/latest/x86_64-centos7-gcc8-opt/setup.sh` this command

**Reply by henryiii** [reply]
The full error is:

```pytb
Traceback (most recent call last):
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/boost_histogram/_internal/axestuple.py", line 101, in __setattr__
    return super().__setattr__(attr, values)
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/hist/axestuple.py", line 63, in name
    for ax, val in zip(self, values, strict=True):
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/hist/axestuple.py", line 17, in zip
    return more_itertools.zip_equal(*iterables)
AttributeError: module 'more_itertools' has no attribute 'zip_equal'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tmpl.py", line 17, in <module>
    h.axes.name = ("x", "y")
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/boost_histogram/_internal/axestuple.py", line 103, in __setattr__
    self.__class__(s.__setattr__(attr, v) for s, v in zip_strict(self, values))
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/boost_histogram/_internal/axestuple.py", line 103, in <genexpr>
    self.__class__(s.__setattr__(attr, v) for s, v in zip_strict(self, values))
  File "/cvmfs/sft-nightlies.cern.ch/lcg/views/dev4cuda/Thu/x86_64-centos7-gcc8-opt/lib/python3.8/site-packages/boost_histogram/_internal/axis.py", line 63, in __setattr__
    object.__setattr__(self, attr, value)
AttributeError: can't set attribute
```

The important error is the first one, not the second one. The problem is that CVMFS has an old more_itertools that doesn't have the zip_equal function. We could either ask CVMFS to update, or replace more_itertools with the hand-rolled implementation in boost-histogram. If they updated to Python 3.10 that would also fix it. :)

**Reply by henryiii** [reply]
I have just released Hist 2.5.2 which removes the need for more-itertools, using the same workaround present in boost-histogram instead. I've reached out to someone with LCG to see if either more-itertools or hist can be bumped.
